packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1589,40 +1596,40 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
     public process(messageArg: ISequencedDocumentMessage, local: boolean) {
         this.verifyNotClosed();
 
-        // If it's not message for runtime, bail out right away.
-        if (!isUnpackedRuntimeMessage(messageArg)) {
-            return;
-        }
-
-        if (this.mc.config.getBoolean("enableOfflineLoad") ?? this.runtimeOptions.enableOfflineLoad) {
-            this.savedOps.push(messageArg);
-        }
-
         // Do shallow copy of message, as methods below will modify it.
         // There might be multiple container instances receiving same message
         // We do not need to make deep copy, as each layer will just replace message.content itself,
         // but would not modify contents details
         let message = { ...messageArg };
 
+        // back-compat: ADO #1385: eventually should become unconditional, but only for runtime messages!
+        // System message may have no contents, or in some cases (mostly for back-compat) they may have actual objects.
+        // Old ops may contain empty string (I assume noops).
+        if (typeof message.contents === "string" && message.contents !== "") {
+            message.contents = JSON.parse(message.contents);
+        }
+
+        // Caveat: This will return false for runtime message in very old format, that are used in snapshot tests
+        // This format was not shipped to production workflows.
+        const runtimeMessage = unpackRuntimeMessage(message);
+
+        if (this.mc.config.getBoolean("enableOfflineLoad") ?? this.runtimeOptions.enableOfflineLoad) {
+            this.savedOps.push(messageArg);
```

Comment:
>messageArg [](http://example.com/codeflow?start=31&length=10)

do we need to push messageArg or message to the saveOps list ?

Reply:
messageArg is the right thing, as we push all ops (not just runtime ops). Take a look at processSavedOps() - it feeds these same ops back to process(), so whatever we get here is what we want to save

This change prepares code for receiving unprocessed ops, but no change is done in loader layer to stop processing ops.
That change will be done in next. I need at least one release to ensure that new loader can work with "old" runtime (where "old" is this PR).

Reply:
> Take a look at processSavedOps() - it feeds these same ops back to process(),

Yes, I had seen and misunderstood it - now it makes sense.